### PR TITLE
key_bundle_cache: add a key bundle cache

### PR DESCRIPTION
**Pull request by aalness**

### CODE REVIEW

File: libkbfs/key_bundle_cache.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// GetTLFReaderKeyBundle implements the KeyBundleCache interface for KeyBundleCacheStandard.
+func (k *KeyBundleCacheStandard) GetTLFReaderKeyBundle(
+	tlf TlfID, bundleID TLFReaderKeyBundleID) *TLFReaderKeyBundleV3 {
+	cacheKey := keyBundleCacheKey{tlf, bundleID.String(), false}
+	if entry, ok := k.lru.Get(cacheKey); ok {
+		if rkb, ok := entry.(*TLFReaderKeyBundleV3); ok {
+			return rkb
+		}
+		// Shouldn't be possible.
+		panic("Invalid key bundle type")
```
Comment (jzila) [inline libkbfs/key_bundle_cache.go]: it's possible that in the future the LRU cache will store entries other than V3 key bundles, right?

File: libkbfs/key_bundle_cache.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// GetTLFWriterKeyBundle implements the KeyBundleCache interface for KeyBundleCacheStandard.
+func (k *KeyBundleCacheStandard) GetTLFWriterKeyBundle(
+	tlf TlfID, bundleID TLFWriterKeyBundleID) *TLFWriterKeyBundleV3 {
+	cacheKey := keyBundleCacheKey{tlf, bundleID.String(), true}
+	if entry, ok := k.lru.Get(cacheKey); ok {
+		if rkb, ok := entry.(*TLFWriterKeyBundleV3); ok {
+			return rkb
+		}
+		// Shouldn't be possible.
+		panic("Invalid key bundle type")
```
Comment (jzila) [inline libkbfs/key_bundle_cache.go]: same comment as above re: potential storage of other types in the cache

File: libkbfs/key_bundle_cache_measured.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+// GetTLFReaderKeyBundle implements the KeyBundleCache interface for
+// KeyBundleCacheMeasured.
+func (b KeyBundleCacheMeasured) GetTLFReaderKeyBundle(
+	tlfID TlfID, bundleID TLFReaderKeyBundleID) (rkb *TLFReaderKeyBundleV3) {
+	b.getReaderBundleTimer.Time(func() {
+		rkb = b.delegate.GetTLFReaderKeyBundle(tlfID, bundleID)
+	})
+	if rkb != nil {
+		b.hitReaderBundleCountMeter.Mark(1)
```
Comment (jzila) [inline libkbfs/key_bundle_cache_measured.go]: is there a value to counting hits if we aren't also counting total number of attempts?

File: libkbfs/key_bundle_cache.go (hunk 1 above)
Comment (aalness) [inline libkbfs/key_bundle_cache.go]: I suppose so. I'm returning an error now instead.

**Comment by aalness** [comment]
>Can key bundles for a given keyBundleCacheKey change?

No. Part of the key is the bundle ID which is the content hash of the bundle.

**Comment by aalness** [comment]
Ready for another pass

File: libkbfs/key_bundle_cache_test.go [hunk 4]
```diff
@@ -37,13 +37,25 @@ func TestKeyBundleCacheBasic(t *testing.T) {
 
 	cache := NewKeyBundleCacheStandard(4)
 
-	if checkWkb := cache.GetTLFWriterKeyBundle(tlf, wkbID); checkWkb != nil {
+	checkWkb, err := cache.GetTLFWriterKeyBundle(tlf, wkbID)
+	if err != nil {
```
Comment (jzila) [inline libkbfs/key_bundle_cache_test.go]: All these checks would be much cleaner using github.com/stretchr/testify/require.go (`require.NoError(t, err)`, `require.NotNil(t, checkWkb)`)

**Comment by aalness** [comment]
Tests cleaned up. Think this is ready whenever

File: libkbfs/key_bundle_cache_test.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+	cache.PutTLFReaderKeyBundle(tlf, rkbID, rkb)
+	// add the same bundle twice
+	cache.PutTLFReaderKeyBundle(tlf, rkbID, rkb)
+	cache.PutTLFReaderKeyBundle(tlf2, rkbID2, rkb2)
+
+	checkWkb, err = cache.GetTLFWriterKeyBundle(tlf, wkbID)
+	require.NoError(t, err)
+	require.NotNil(t, checkWkb)
+	require.True(t, reflect.DeepEqual(checkWkb, wkb))
```
Comment (jzila) [inline libkbfs/key_bundle_cache_test.go]: FWIW `require.Equal(t, chekWkb, wkb)` does a `reflect.DeepEqual`

**Pull request merged**